src/relations/peers.py

```diff
@@ -0,0 +1,255 @@
+│ unit data        │ ╭─ pgbouncer-k8s/0 ─╮ ╭─ pgbouncer-k8s/1* ─╮                                                     │
+│                  │ │ <empty>           │ │ <empty>            │                                                     │
+│                  │ ╰───────────────────╯ ╰────────────────────╯                                                     │
+└──────────────────┴──────────────────────────────────────────────────────────────────────────────────────────────────┘
+
+"""  # noqa: W505
+
+import logging
+from typing import Optional
+
+from charms.pgbouncer_k8s.v0.pgb import PgbConfig
+from ops.charm import CharmBase, RelationChangedEvent, RelationCreatedEvent
+from ops.framework import Object
+from ops.pebble import ConnectionError
+from constants import PEER_RELATION_NAME
+
+CFG_FILE_DATABAG_KEY = "cfg_file"
+AUTH_FILE_DATABAG_KEY = "auth_file"
+
+
+logger = logging.getLogger(__name__)
+
+
+class Peers(Object):
+    """Defines functionality for the pgbouncer peer relation.
+
+    The data created in this relation allows the pgbouncer charm to connect to the postgres charm.
+
+    Hook events observed:
+        - relation-created
+        - relation-joined
+        - relation-changed
+    """
+
+    def __init__(self, charm: CharmBase):
+        super().__init__(charm, PEER_RELATION_NAME)
+
+        self.charm = charm
+
+        self.framework.observe(charm.on[PEER_RELATION_NAME].relation_created, self._on_created)
+        self.framework.observe(charm.on[PEER_RELATION_NAME].relation_joined, self._on_changed)
+        self.framework.observe(charm.on[PEER_RELATION_NAME].relation_changed, self._on_changed)
+
+    @property
+    def app_databag(self):
+        """Returns the app databag for the Peer relation."""
+        peer_relation = self.model.get_relation(PEER_RELATION_NAME)
+        if peer_relation is None:
+            return {}
+        return peer_relation.data[self.charm.app]
+
+    @property
+    def unit_databag(self):
+        """Returns the app databag for the Peer relation."""
+        peer_relation = self.model.get_relation(PEER_RELATION_NAME)
+        if peer_relation is None:
+            return {}
+        return peer_relation.data[self.charm.unit]
+
+    def _on_created(self, event: RelationCreatedEvent):
+        if not self.charm.unit.is_leader():
+            return
+
+        try:
+            cfg = self.charm.read_pgb_config()
+        except FileNotFoundError:
+            # If there's no config, the charm start hook hasn't fired yet, so defer until it's
+            # available.
+            event.defer()
+            return
+
+        self.update_cfg(cfg)
+
+        try:
+            if self.charm.backend:
+                # update userlist only if backend relation exists.
+                self.update_auth_file(self.charm.read_auth_file())
+        except FileNotFoundError:
+            # Auth files will be added to the databag once they're available
```

Comment:
Don't you need to defer or trigger the event again in some place to populate the `AUTH_FILE_DATABAG_KEY` in the databag? Sorry if I am missing it in another place (I checked some calls and couldn't find it).

Reply:
when the auth_file is added it should create a relation_changed event. This is called in render_auth_file, in charm.py line 290. If anything, the config deferral should be unnecessary - I'm testing that now.

Reply:
Thanks for the clarification, Will.